R-packages/modeltools/R/response.R

```diff
@@ -0,0 +1,95 @@
+library(assertthat)
+
+#' Compute the corresponding observed response values for a set of signals.
+#'
+#' The response for an observed data point at time `t` is defined as the mean of observed values on
+#' days `t + s` through `t + e` (inclusive) multiplied by `e - s`, where `s` and `e` are the number
+#' of days from `forecast_date` to the start and end, respectively, of the next full
+#' `incidence_period`.  We choose to take the mean and multiply by the incidence period length
+#' rather than simply sum the observed values in order to avoid treating missing values as 0.
+#'
+#' @param signals Wide data frame of signals.  We expect the following columns:
+#'     \itemize{
+#'     \item{`geo_value`}{Strings of geographic locations.}
+#'     \item{`time_value`}{Dates of training data.}
+#'     \item{Covariate columns}{Columns with names of the form `value-{days}:{signal}` or
+#'         `value+0:{signal} whose values correspond to `{signal}` `{days}` before `time_value`}
+#'     }
+#' @param response_name Name of response signal.  The value of the response variable on a single
+#'     day should be located in column `value+0:{response_name}` of `signals`.
+#' @param forecast_date Date on which the forecast will be made
+#' @param incidence_period Time period over which the response should be summed.
+#' @param ahead integer or vector of integer ahead values
+#'
+#' @return Data frame of signals equal to `signals` with `length(ahead)` additional columns with
+#'     names `response+{a}:{response_name}` corresponding to the response variable at ahead `a`. 
+#'
+#' @examples
+#' \dontrun{
+#' add_response_columns(
+#'   tibble(
+#'     geo_value = "mi"
+#'     time_value = seq.Date(as.Date("2020-07-01"), as.Date("2020-07-14"), "day"),
+#'     `value+0:sig` = 1:14
+#'   ),
+#'   "sig",
+#'   as.Date("2020-07-15"),
+#'   "epiweek",
+#'   1:2 
+#' )
+#' }
+#'
+#' @export
+add_response_columns <- function(signals,
+                                 response_name,
+                                 forecast_date,
+                                 incidence_period,
+                                 ahead) {
+    out_df <- signals
+    for (a in ahead) {
+        target_period <- as.list(get_target_period(forecast_date, incidence_period, a)[1, ])
+        window_size <- as.numeric(target_period$end - target_period$start) + 1
+        out_df <- slide_by_geo(out_df,
+                               ~ Mean(window_size * .x[[paste0("value+0:", response_name)]]),
+                               n = window_size,
+                               shift = as.numeric(target_period$end - forecast_date),
+                               col_name = paste0("response+", a, ":", response_name))
+    }
+    return(out_df)
+}
+
+#' Get the target period for a forecast date, incidence period and ahead
+#'
+#' 
+get_target_period <- function(forecast_date, incidence_period, ahead) {
```

Comment:
This is in `evalcast`. I don't mind duplicating here to try to avoid dependence if that's the goal.

Reply:
I would prefer to just call the `evalcast` version but this is not currently possible because the most recent `evalcast` version is not in the `main` or `modeltools` branches.

Reply:
I think `evalcast` is already in the dependencies, so maybe better just to import.

Reply:
Ugh, you're right. The dependency is on `main`. Ok, forget my comment.